File: module-4-filesystem/solutions/python/sandbox.py
"""Module 4: Path sandbox."""
import os


class Sandbox:
    def __init__(self, root: str):
        self.root = os.path.realpath(root)

    def resolve(self, file_path: str) -> str:
        resolved = os.path.realpath(os.path.join(self.root, file_path))
        if not resolved.startswith(self.root):
            raise ValueError(f'Path "{file_path}" escapes sandbox root "{self.root}"')
        return resolved

File: module-4-filesystem/solutions/sandbox.ts
/**
 * Module 4: Path sandbox
 * All file operations must resolve under the allowed root.
 */
import * as path from "path";

export class Sandbox {
  constructor(public readonly root: string) {}

  resolve(filePath: string): string {
    const resolved = path.resolve(this.root, filePath);
    if (!resolved.startsWith(this.root)) {
      throw new Error(`Path "${filePath}" escapes sandbox root "${this.root}"`);
    }
    return resolved;
  }
}
